ibis/backends/dask/execution/reductions.py

```diff
@@ -73,37 +74,31 @@ def dask_execute_node_least_list(op, value, **kwargs):
     return compute_row_reduction(da.minimum, value)
 
 
-# TODO - aggregations - #2553
 @execute_node.register(ops.Reduction, ddgb.SeriesGroupBy, type(None))
 def execute_reduction_series_groupby(
     op, data, mask, aggcontext=None, **kwargs
 ):
     return aggcontext.agg(data, type(op).__name__.lower())
 
 
-def _filtered_reduction(mask, method, data):
-    return method(data[mask[data.index]])
+def _filtered_reduction(data, mask):
+    return make_selected_obj(data)[mask.obj].groupby(data.index)
```

Comment:
The original pandas version doesn't have groupby, so I am a bit confused what this method is supposed to do.

Reply:
The pandas version has different things passed to it. It is called via `data, functools.partial(_filtered_reduction, mask.obj, method)` where `mask.obj` gives you the underlying ungrouped Series. This partial then gets applied in aggctx. In the dask version, we regroup so we can call `.sum` (or whatever) directly. They perform the same functionality, just through a different method.

Reply:
I see. I am curious it's reason that we don't follow pandas structure to pass a dd.SeriesGroupBy here?

Reply:
So there are few things at work in the pandas backend. Looking that the following code

```python
def _filtered_reduction(mask, method, data):
    return method(data[mask[data.index]])


@execute_node.register(ops.Reduction, SeriesGroupBy, SeriesGroupBy)
def execute_reduction_series_gb_mask(
    op, data, mask, aggcontext=None, **kwargs
):
    method = operator.methodcaller(type(op).__name__.lower())
    return aggcontext.agg(
        data, functools.partial(_filtered_reduction, mask.obj, method)
    )
``` 
via https://github.com/ibis-project/ibis/blob/master/ibis/backends/pandas/execution/generic.py#L526

Here `aggcontext.agg` will get passed a partial that will perform the filtering and then apply the reduction method. This will get called in https://github.com/ibis-project/ibis/blob/master/ibis/backends/pandas/aggcontext.py#L353, specifically via the callable branch. 

Dask does not play well with arbitrary user defined aggregations. You have to use https://docs.dask.org/en/latest/dataframe-api.html#custom-aggregation to create your own wrapper. If you pass it a partial it will not know how to work with that. Luckily dask already defines aggregations for all the common operations we have in `type(op).__name__.lower()`. So we construct the proper grouped and filtered Series and dispatch directly to dask for the actual aggregation via https://github.com/ibis-project/ibis/blob/master/ibis/backends/pandas/aggcontext.py#L355